### ISSUE TITLE
World Anchor Store never ready

### ISSUE THREAD

**Issue by Reeley** [Microsoft/HoloToolkit-Unity#640]
I stumbled upon this problem yesterday, when i first got into world anchors. Basically what i tried is to get the World Anchor Manager working, but the problem i had is that the callback when the anchor store is ready never got called. So i thought maybe i just messed up and created a new project with just the spatialmapping and world anchor store manager in it. Well the end of the story was that it still didnt work.

So i fiddled around a little bit and changed:
```
protected override void Awake()
        {
            base.Awake();

            AnchorStore = null;
            WorldAnchorStore.GetAsync(AnchorStoreReady);
        }
```

to

```
void Start()
        {
            AnchorStore = null;
            WorldAnchorStore.GetAsync(AnchorStoreReady);
        }
```

Basically i let the code in the Awake() function run in the Start() function. Maybe there is something not yet initialized when Awake() gets called, cause i also got an NullReference Error.

Is this a bug? Should the code in Awake() run in the Start() function?

**Comment by Nakda** [comment]
The WorldAnchorManager should work just fine as it is.

As far as I know, the WorldAnchorStore callback (AnchorStoreReady here) is never fired within the Unity Editor. You have to deploy to HoloLens.

What is the NullReferenceException about ?

**Comment by Reeley** [comment]
this is the all info when debugging with vs:

```
"WorldAnchor.exe" (CoreCLR: CoreCLR_UWP_Domain): "C:\Data\Users\DefaultAccount\AppData\Local\DevelopmentFiles\WorldAnchorVS.Release_x86.l.heft\System.Linq.Expressions.dll" geladen. Das Laden von Symbolen wurde übersprungen. Das Modul ist optimiert, und die Debugoption "Nur eigenen Code" ist aktiviert.
trying to load anchor store
 
(Filename: C:/buildslave/unity/build/artifacts/generated/Metro/runtime/DebugBindings.gen.cpp Line: 51)


loading anchor store...
 
(Filename: C:/buildslave/unity/build/artifacts/generated/Metro/runtime/DebugBindings.gen.cpp Line: 51)


"WorldAnchor.exe" (CoreCLR: CoreCLR_UWP_Domain): "C:\Data\Users\DefaultAccount\AppData\Local\DevelopmentFiles\WorldAnchorVS.Release_x86.l.heft\System.Diagnostics.StackTrace.dll" geladen. Das Laden von Symbolen wurde übersprungen. Das Modul ist optimiert, und die Debugoption "Nur eigenen Code" ist aktiviert.
Starting the observer.
 
(Filename: C:/buildslave/unity/build/artifacts/generated/Metro/runtime/DebugBindings.gen.cpp Line: 51)


Exception: Incorrect hashcode or refCount - (GCHandleToObject)
Hash: expected 45867655 was 21358135(HashCode: 21358135 BlockID: 0 HandleIdInBlock: 259), RefCount: 2
(HashCode: 21358135 BlockID: 0 HandleIdInBlock: 259), AllocatedObjectCount: 25
MainThreadId: 4, ThisThreadId: 4
StackTrace:    at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at WinRTBridge.GCHandledObjects.ValidateObject(ObjectDebugInfo dbgInfo, Object o, String message)
   at WinRTBridge.GCHandledObjects.GCHandleToObject(Int64 handleId)
   at UnityEngine.VR.WSA.Persistence.WorldAnchorStore.$Invoke12(Int64 instance, Int64* args)
   at UnityEngine.Internal.$MethodUtility.InvokeMethod(Int64 instance, Int64* args, IntPtr method)
Type: System.Exception
Module: 
InnerException: <No Data>
AdditionalInfo:<No Data>
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at WinRTBridge.ExceptionHandling.LogException(Exception ex, String additionalInfo)
   at WinRTBridge.GCHandledObjects.ValidateObject(ObjectDebugInfo dbgInfo, Object o, String message)
   at WinRTBridge.GCHandledObjects.GCHandleToObject(Int64 handleId)
   at UnityEngine.VR.WSA.Persistence.WorldAnchorStore.$Invoke12(Int64 instance, Int64* args)
   at UnityEngine.Internal.$MethodUtility.InvokeMethod(Int64 instance, Int64* args, IntPtr method)
 
(Filename:  Line: 0)


InvalidCastException: Unable to cast object of type 'UnityEngine.GameObject' to type 'GetAsyncDelegate'.
   at UnityEngine.VR.WSA.Persistence.WorldAnchorStore.$Invoke12(Int64 instance, Int64* args)
   at UnityEngine.Internal.$MethodUtility.InvokeMethod(Int64 instance, Int64* args, IntPtr method) 
(Filename: <Unknown> Line: 0)
```

**Comment by Nakda** [comment]
I don't see the NullReferenceException you were mentioning initially. 

How did you check if the AnchorStoreReady callback is ever called ? Could you share some of your code so I can see what's going on ?

**Comment by Reeley** [comment]
Basically im using the standard WorldAnchorManager script which is included in the HoloToolkit. I just added some debug logs:

```
/// <summary>
        /// Callback function that contains the WorldAnchorStore object.
        /// </summary>
        /// <param name="anchorStore">The WorldAnchorStore to cache.</param>
        private void AnchorStoreReady(WorldAnchorStore anchorStore)
        {
            print("Anchor store ready");
            AnchorStore = anchorStore;
        }

        /// <summary>
        /// When the app starts grab the anchor store immediately.
        /// </summary>
        protected override void Awake()
        {
            base.Awake();

            AnchorStore = null;
            print("trying to load anchor store");
            WorldAnchorStore.GetAsync(AnchorStoreReady);
            print("loading anchor store...");
        }
```

i dont know where the nullreference went, i had to setup the project again and it disappeared.

> I just had the same! I am using the ImportExportAnchorManager. The callback to AnchorStoreReady was never called. Finally moved the GetAsync call to Start and no everything works as expected!

Glad it helped you :)

**Comment by christjt** [comment]
I second this. Had to change the GetAsync call from awake to start. The exception only first came when switching to Unity 5.6 however...

**Comment by Nakda** [comment]
You might want to contribute to this thread:
https://forum.unity3d.com/threads/error-running-in-hololens.449303/

**Comment by FejZa** [comment]
Can confirm this. My world anchor store was not working becasue GetAsync never invoked the ready callback. "Fixed" it by moving the call to GetAsync to the Start method instead of Awake.

**Comment by Cameron-Micka** [comment]
I can also confirm this became a problem for us after upgrading to Unity 5.6. Also, "Fixed" by moving the call to GetAsync to the Start method instead of Awake.

**Comment by AIproj** [comment]
Can confirm too. Moving GetAsync to start fixed it, although it is not necessarily ideal since scripts such as TapToPlace.cs may rely on the anchorstore being ready when Start() is called. A temporary solution is to delay the execution of such scripts in the meanwhile.

**Comment by Ziugy** [comment]
It has been helpful for me to add an event on `WorldAnchorManager` that gets fired when `WorldAnchorStore.GetAsync` completes.

**Comment by StephenHodgson** [comment]
@Ziugy has suggested the best solution in my opinion.

**Comment by AR4DS** [comment]
void Start()
    {
        WorldAnchorStore.GetAsync(myAsyncDelegate);
        text.text = store.anchorCount.ToString();//NullReferenceException
    }

    public void myAsyncDelegate(WorldAnchorStore store){
        this.store = store;
    }
always failed let me lost..

**Comment by StephenHodgson** [comment]
@AR4DS, the `text.text = store.anchorCount.ToString();` will almost always be null because you're trying to access the store before the async operation has completed.  Try a coroutine.

**Comment by AR4DS** [comment]
@StephenHodgson i've used coroutine to wait 10's, but still doesn't solve, maybe Editor/Enumerator cause it.🙃

**Comment by AR4DS** [comment]
@StephenHodgson coroutine still can't resolve it, maybe that's special case that using editor and enumerator.

**Comment by Reeley** [comment]
If you have problems or questions please post them here [HoloLens Forum](https://forums.hololens.com/categories/questions-and-answers).

Try this:

```
void Start()
{
    WorldAnchorStore.GetAsync(myAsyncDelegate);
}

public void myAsyncDelegate(WorldAnchorStore store){
    this.store = store;
    text.text = store.anchorCount.ToString();
}
```

**Comment by StephenHodgson** [comment]
https://issuetracker.unity3d.com/issues/wsa-worldanchorstore-dot-getasync-throws-an-exception-when-starting-deployed-hololens-app-built

**Comment by AR4DS** [comment]
@StephenHodgson that's my code, store init succeed, but anchors can't add to the store when i call tap event(
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.VR.WSA.Input;
using UnityEngine.VR.WSA;
using UnityEngine.VR.WSA.Persistence;
using UnityEngine.UI;

public class Test : MonoBehaviour
{
    public Transform prefabTran;
    public Text text;
    List<WorldAnchor> was = new List<WorldAnchor>();
    WorldAnchorStore store;
    GestureRecognizer gestureR;

    void Start()
    {
        InitGesture();
        StartCoroutine(WorldStoreInstance());
    }

    void InitGesture()
    {
        gestureR = new GestureRecognizer();
        gestureR.SetRecognizableGestures(GestureSettings.Tap);
        gestureR.TappedEvent += myTapEvent;
        gestureR.StartCapturingGestures();
    }
    
    private void myTapEvent(InteractionSourceKind source, int tapCount, Ray headRay)
    {
        UpdateInfo();
    }

    int i = 0;
    void UpdateInfo()
    {
        InitWAStore();
        text.text = ++i + "Count = " + store.anchorCount + " in " + was.Count + "\n";
        string[] ids = store.GetAllIds();
        foreach (string id in ids)
        {
            text.text += id + "\t";
        }
    }

    IEnumerator WorldStoreInstance()
    {
        WorldAnchorStore.GetAsync(myAsyncDelegate);
        yield return store;
        CreateSpheres();
    }

    void CreateSpheres()
    {
        for (int x = -1; x < 2; x++)
        {
            for (int y = -1; y < 2; y++)
            {
                for (int z = 10; z < 13; z++)
                {
                    Transform t = Instantiate(prefabTran, new Vector3(x, y, z), Quaternion.identity);
                    t.name = x.ToString() + y.ToString() + z.ToString();
                    was.Add(t.GetComponent<WorldAnchor>());
                }
            }
        }
    }

    void InitWAStore()
    {
        for(int i = 0; i < was.Count; i++)
        {
            store.Save(was[i].gameObject.name, was[i]);
        }
    }

    public void myAsyncDelegate(WorldAnchorStore store){
        this.store = store;
    }
}

**Comment by AR4DS** [comment]
@StephenHodgson that's my code, store init succeed, but anchors can't add to the store when i call tap event(
`using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.VR.WSA.Input;
using UnityEngine.VR.WSA;
using UnityEngine.VR.WSA.Persistence;
using UnityEngine.UI;

public class Test : MonoBehaviour
{
    public Transform prefabTran;
    public Text text;
    List<WorldAnchor> was = new List<WorldAnchor>();
    WorldAnchorStore store;
    GestureRecognizer gestureR;

    void Start()
    {
        InitGesture();
        StartCoroutine(WorldStoreInstance());
    }

    void InitGesture()
    {
        gestureR = new GestureRecognizer();
        gestureR.SetRecognizableGestures(GestureSettings.Tap);
        gestureR.TappedEvent += myTapEvent;
        gestureR.StartCapturingGestures();
    }
    
    private void myTapEvent(InteractionSourceKind source, int tapCount, Ray headRay)
    {
        UpdateInfo();
    }

    int i = 0;
    void UpdateInfo()
    {
        InitWAStore();
        text.text = ++i + "Count = " + store.anchorCount + " in " + was.Count + "\n";
        string[] ids = store.GetAllIds();
        foreach (string id in ids)
        {
            text.text += id + "\t";
        }
    }

    IEnumerator WorldStoreInstance()
    {
        WorldAnchorStore.GetAsync(myAsyncDelegate);
        yield return store;
        CreateSpheres();
    }

    void CreateSpheres()
    {
        for (int x = -1; x < 2; x++)
        {
            for (int y = -1; y < 2; y++)
            {
                for (int z = 10; z < 13; z++)
                {
                    Transform t = Instantiate(prefabTran, new Vector3(x, y, z), Quaternion.identity);
                    t.name = x.ToString() + y.ToString() + z.ToString();
                    was.Add(t.GetComponent<WorldAnchor>());
                }
            }
        }
    }

    void InitWAStore()
    {
        for(int i = 0; i < was.Count; i++)
        {
            store.Save(was[i].gameObject.name, was[i]);
        }
    }

    public void myAsyncDelegate(WorldAnchorStore store){
        this.store = store;
    }
}
`

**Comment by AR4DS** [comment]
thx @Reeley, it works!

**Comment by edwinckc** [comment]
Did anyone try Unity 5.6.1p4, and did it fix this problem for you?
I'm using it now, and it fixed [this other problem](https://forum.unity3d.com/threads/error-running-in-hololens.449303/), but not the WorldAnchorStore.GetAsync(callback) problem where the anchor store is never ready.

**Comment by Reeley** [comment]
In the changelogs for Unity 5.6.2 they say they fixed issue 902234 (Windows Store: Fixed WorldAnchorStore.GetAsync crashing the player. (902234))

can someone test if this fixed the bug? (i dont have a hololens currently so i cant test :/ )

**Comment by StephenHodgson** [comment]
Didn't even know 5.6.2 was out, lol.

The 5.6.1p4 definitely fixed the issue with the world anchor store.

**Comment by Reeley** [comment]
awesome! :D

**Comment by hammadshah38** [comment]
Hi everyone! I know it's too late to comment on this thread but I need your help regarding world anchors in hololens.
Does world anchor manager GetAsync() call automatically position the objects based on saved anchors?

Waiting for your kind response.

Thank you and regards
